proposals/003-metric-improvenents.md

```diff
@@ -0,0 +1,262 @@
+absent.
+
+`originating_filter` is only used for requests sent by the filters themselves (i.e. the `sendRequest` API)
+and their responses.  It will be populated with the filter's name.  Otherwise `originating_filter` will be absent.
+
+### New distribution `kroxylicious_message_process_time`
+
+This metric records the length of time a message (request or response) has taken to traverse the proxy.
+
+* The start time will be time the bytes containing the message arrived from the network (in the Netty handler, before any decoding).
+* The end time will be the time that the network write at the proxy's opposite end completes (i.e. the Netty write promise completes)
+
+The use cases supported by this metric are ones where you are interested in how much processing time it being incurred 
+by the proxy decoding and encoding messages and any processing time incurred by the filter.
+
+The requests sent by filter themselves, the start time for a request will be time filter called `#sendRequest`. The end time
+will be the time the network write at the proxy upstream completes.  For responses, the start time is time the bytes containing
+the message arrived from the network. The end time will be the time the response reaches the originating filter (i.e. the future completes).
+
+### New gauges `kroxylicious_(up|down)stream_backpressure_state`
+
+Reports the backpressure state for upstream and downstream.  The use-case for metric is to allow the user to
+understand when the proxy must pause reading owing whilst asynchronous work completes (for example, in record validation
+requesting a schema or in record encryption, generating a new DEK).
+
+This gauge will use values `0` and `1` only.  `1` will signify that the proxy has stopped reading the channel whilst
+it awaits the completion of asynchronous work.  `0` signifies that there is no pending asynchronous.
+
+### New distributions `kroxylicious_(up|down)stream_backpressure_time`
+
+Reports the total length of time backpressure has been exerted.
+
+### Add label `node_id` and `virtual_cluster` to `kroxylicious_(down|up)stream_connections_attempts` and `kroxylicious_(down|up)stream_errors`
+
+`node_id` will only be present if for requests going to or responses coming from a broker.  For bootstrap interactions,
+`node_id` will be absent.
+
+The existing label `virtualCluster` will be deprecated.  A new label `virtual_cluster` will be added.  It will carry
+the virtual cluster's name.
+
+### Use `kroxylicious_downstream_errors` use to record other downstream errors
+
+Issues should as downstream TLS negotiating errors or failure to resolve virtual cluster result in the connection being
+closed, however, there is no metric counting those errors. `kroxylicious_downstream_connections_attempts` and
+`kroxylicious_downstream_errors` should be incremented for these case too.  `virtual_cluster` and `node_id` won't be known
+so these labels should be omitted.
```

Comment:
```suggestion
Issues such as downstream TLS negotiating errors or failure to resolve virtual cluster result in the connection being
closed, however, there is no metric counting those errors. `kroxylicious_downstream_connections_attempts` and
`kroxylicious_downstream_errors` should be incremented for these case too.  `virtual_cluster` and `node_id` won't be known
so these labels should be omitted.
```

Reply:
Downstream TLS is an awkward case as I'm not sure we have the appropriate call backs or events to track it specifically but I think errors from downstream TLS and the [net filter](https://github.com/kroxylicious/kroxylicious/blob/79dd894bf4b69a184e5cc0dc72570ef8cb6a367d/kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyInitializer.java#L281-L284) bubble up as exceptions and are caught and fed into the existing `kroxylicious_downstream_errors` counter via the ProxyChannelStateMachine. 

Issues where the netFilter selects something we can't connect to should also be fed into an error counter already.

Reply:
> think errors from downstream TLS and the [net filter](https://github.com/kroxylicious/kroxylicious/blob/79dd894bf4b69a184e5cc0dc72570ef8cb6a367d/kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyInitializer.java#L281-L284) bubble up as exceptions and are caught and fed into the existing kroxylicious_downstream_errors counter via the ProxyChannelStateMachine

I don't see that behaviour.  A test has been written to show that's not the case.
https://github.com/kroxylicious/kroxylicious/pull/2230

Reply:
Thanks Keith! Looking at #2230 I realised the error in my thinking 🤔 The initialiser is handling this before the state machine kicks in.